File: lucid/_C/utils/tokenizer/WordPiece.cpp
// lucid/_C/utils/tokenizer/WordPiece.cpp
//
// WordPiece algorithm implementation.  See WordPiece.h for the API
// contract + algorithm summary.  This file is the canonical encode /
// decode / train surface for every WordPiece-based checkpoint Lucid
// loads — the Python ``WordPieceTokenizer`` wrapper holds a
// ``std::unique_ptr<WordPiece>`` and forwards directly through the
// v-table.
//
// Implementation notes
// --------------------
// * Encode is greedy longest-match per word with an O(W^2) inner
//   cost (W = word length in codepoints).  Acceptable because BERT
//   capped word length at 100; we honour the same cap via
//   ``max_chars_per_word_``.
// * Train uses simple whitespace pre-tokenization + greedy frequency
//   pair-merging — a BPE-style approximation of the full likelihood
//   objective, matching the HF ``tokenizers`` crate's
//   ``WordPieceTrainer``.
// * UTF-8 codepoints are honoured during both encode and train so
//   multi-byte characters are never split.
// * Continuation pieces carry the ``##`` prefix as a literal vocab
//   key — encode prepends it on non-leading segments and decode
//   strips it back off.

#include "WordPiece.h"

#include <algorithm>
#include <stdexcept>

namespace lucid::utils::tokenizer {

// Default constructor — empty vocab.  Populate via ``train`` or by
// re-constructing with an explicit vocab from ``vocab.txt``.
WordPiece::WordPiece() = default;

// Construct from explicit ``vocab.txt``-style mapping plus the BERT
// knobs (UNK token, ``##`` continuation prefix, max chars per word).
// ``rebuild_tables_`` then populates the dense reverse table and
// caches ``unk_id_`` so the encode hot path skips a hash lookup.
WordPiece::WordPiece(std::unordered_map<std::string, TokenId> vocab,
                     std::string unk_token,
                     std::string continuing_prefix,
                     std::size_t max_chars_per_word)
    : vocab_(std::move(vocab)),
      unk_token_(std::move(unk_token)),
      continuing_prefix_(std::move(continuing_prefix)),
      max_chars_per_word_(max_chars_per_word) {
    rebuild_tables_();
}

// Re-derive every cached table (``id_to_token_``, ``unk_id_``) from
// the canonical ``vocab_`` map.  Invariant: must be called after ANY
// mutation of ``vocab_`` — encode/decode read the cached tables
// directly and would otherwise see stale data.
void WordPiece::rebuild_tables_() {
    TokenId max_id = -1;
    for (const auto& [tok, id] : vocab_)
        if (id > max_id)
            max_id = id;
    id_to_token_.assign(static_cast<std::size_t>(max_id + 1), std::string{});
    for (const auto& [tok, id] : vocab_) {
        if (id >= 0 && static_cast<std::size_t>(id) < id_to_token_.size())
            id_to_token_[static_cast<std::size_t>(id)] = tok;
    }
    auto it = vocab_.find(unk_token_);
    unk_id_ = (it == vocab_.end()) ? -1 : it->second;
}

// O(1) reverse lookup via the dense ``id_to_token_`` array.  Empty
// string for out-of-range ids (matches the BERT-tokenizers convention
// rather than throwing).
std::string WordPiece::id_to_token(TokenId id) const {
    if (id < 0 || static_cast<std::size_t>(id) >= id_to_token_.size())
        return "";
    return id_to_token_[static_cast<std::size_t>(id)];
}

// Greedy longest-match WordPiece for ONE pre-tokenized word.  Three
// stages — (1) whole-word fast path, (2) BERT's max-length guard
// (default 100 chars → UNK), (3) left-to-right scan that picks the
// longest vocab-resident prefix at each position, applying the ``##``
// continuation prefix to every non-leading segment.  If no valid
// prefix exists at any position the whole word collapses to a single
// UNK (BERT semantics).  All segmentation seeks UTF-8 codepoint
// boundaries so multi-byte characters survive intact.
IdSequence WordPiece::encode_word_(const std::string& word) const {
    // 1. Whole-word match shortcut.
    if (auto it = vocab_.find(word); it != vocab_.end())
        return {it->second};
    // 2. Cap on length — BERT drops anything longer than 100 chars.
    if (word.size() > max_chars_per_word_) {
        if (unk_id_ >= 0)
            return {unk_id_};
        return {};
    }
    // 3. Greedy longest-match left-to-right.  Walk through UTF-8
    //    codepoint boundaries so we don't split a multi-byte char.
    //    Pre-compute the codepoint start offsets for O(1) seek.
    std::vector<std::size_t> cp_offsets;
    cp_offsets.reserve(word.size() + 1);
    std::size_t i = 0;
    while (i < word.size()) {
        cp_offsets.push_back(i);
        unsigned char c0 = static_cast<unsigned char>(word[i]);
        std::size_t cp_len;
        if (c0 < 0x80)
            cp_len = 1;
        else if ((c0 >> 5) == 0b110)
            cp_len = 2;
        else if ((c0 >> 4) == 0b1110)
            cp_len = 3;
        else if ((c0 >> 3) == 0b11110)
            cp_len = 4;
        else
            cp_len = 1;
        if (i + cp_len > word.size())
            cp_len = word.size() - i;
        i += cp_len;
    }
    cp_offsets.push_back(word.size());

    IdSequence ids;
    std::size_t cp_start = 0;  // index into cp_offsets
    while (cp_start + 1 < cp_offsets.size()) {
        // Try longest prefix first.
        TokenId match_id = -1;
        std::size_t match_cp_end = cp_start;
        for (std::size_t cp_end = cp_offsets.size() - 1; cp_end > cp_start; --cp_end) {
            std::string sub =
                word.substr(cp_offsets[cp_start], cp_offsets[cp_end] - cp_offsets[cp_start]);
            if (cp_start > 0)
                sub = continuing_prefix_ + sub;
            auto it = vocab_.find(sub);
            if (it != vocab_.end()) {
                match_id = it->second;
                match_cp_end = cp_end;
                break;
            }
        }
        if (match_id < 0) {
            // No valid prefix at this position → entire word is UNK.
            if (unk_id_ >= 0)
                return {unk_id_};
            return {};
        }
        ids.push_back(match_id);
        cp_start = match_cp_end;
    }
    return ids;
}

// Public encode entry point.  Splits on whitespace as a fallback when
// called directly; the Python wrapper normally pre-tokenizes via the
// BertNormalizer + WhitespacePunctuationSplit chain before invoking.
IdSequence WordPiece::encode(const std::string& text) const {
    // The Python wrapper splits the input into pre-tokenized words
    // before calling encode (via the configured WordPieceNormalizer
    // + WhitespacePunctuationSplit chain).  If ``encode`` is called
    // directly on a multi-word string, we fall back to a simple
    // whitespace split here.
    IdSequence out;
    std::size_t i = 0;
    while (i < text.size()) {
        while (i < text.size() &&
               (text[i] == ' ' || text[i] == '\t' || text[i] == '\n' || text[i] == '\r'))
            ++i;
        if (i >= text.size())
            break;
        std::size_t start = i;
        while (i < text.size() &&
               !(text[i] == ' ' || text[i] == '\t' || text[i] == '\n' || text[i] == '\r'))
            ++i;
        auto word_ids = encode_word_(text.substr(start, i - start));
        out.insert(out.end(), word_ids.begin(), word_ids.end());
    }
    return out;
}

// Decode by stitching surface forms back together.  Continuation
// pieces (``##`` prefix) glue onto the previous token without a
// space; every non-continuation token gets a leading space (except
// the very first).  Out-of-range ids are skipped silently.
std::string WordPiece::decode(const IdSequence& ids) const {
    // Join tokens; strip leading "##" from continuation pieces and
    // emit a space before each non-continuation token.
    std::string out;
    bool first = true;
    for (TokenId id : ids) {
        if (id < 0 || static_cast<std::size_t>(id) >= id_to_token_.size())
            continue;
        const std::string& tok = id_to_token_[static_cast<std::size_t>(id)];
        bool is_continuation = tok.size() >= continuing_prefix_.size() &&
                               tok.compare(0, continuing_prefix_.size(), continuing_prefix_) == 0;
        if (is_continuation) {
            out += tok.substr(continuing_prefix_.size());
        } else {
            if (!first)
                out += ' ';
            out += tok;
        }
        first = false;
    }
    return out;
}

// ── Training ────────────────────────────────────────────────────────
//
// Greedy frequency-based training (approximates HF's
// WordPieceTrainer).  Algorithm:
//
//   1. Build word frequency dict from corpus via whitespace split.
//   2. Seed vocab with every distinct char + continuation forms
//      ``##char`` for every non-leading char.
//   3. Iteratively pick the most-frequent (left, right) bigram
//      across all current symbol sequences + merge into a new token
//      (the merged token's surface form keeps ``##`` only if BOTH
//      halves were continuations or if the right half was — matches
//      BERT vocab layout).
//
// This is the same logic as BPE training but with the ``##``
// continuation suffix machinery; for our purposes that's sufficient
// to produce a BERT-compatible vocab.

namespace {

struct WordState {
    std::vector<std::string> symbols;
    std::uint64_t count;
};

}  // namespace

// Public training entry point.  Three-phase pipeline — (1) tally
// word frequencies via whitespace pre-tokenization, (2) seed vocab
// with every codepoint (plain + ``##`` continuation) and the UNK,
// (3) greedy bigram-frequency merging until ``target_vocab_size`` is
// reached or no further pair occurs more than once.  Merged surface
// forms strip the right half's ``##`` prefix (keeping the left's, if
// any) so the resulting vocab matches BERT layout byte-for-byte.
// NOT thread-safe — mutates ``vocab_`` / cached tables in place.
void WordPiece::train(const std::vector<std::string>& corpus, std::size_t target_vocab_size) {
    // 1. Word frequencies.
    std::unordered_map<std::string, std::uint64_t> word_freq;
    for (const auto& doc : corpus) {
        std::size_t i = 0;
        while (i < doc.size()) {
            while (i < doc.size() &&
                   (doc[i] == ' ' || doc[i] == '\t' || doc[i] == '\n' || doc[i] == '\r'))
                ++i;
            std::size_t start = i;
            while (i < doc.size() &&
                   !(doc[i] == ' ' || doc[i] == '\t' || doc[i] == '\n' || doc[i] == '\r'))
                ++i;
            if (i > start)
                ++word_freq[doc.substr(start, i - start)];
        }
    }
    // 2. Seed vocab with chars + continuation chars + the UNK token.
    vocab_.clear();
    TokenId next_id = 0;
    if (!unk_token_.empty()) {
        vocab_[unk_token_] = next_id++;
    }
    auto add_token = [&](const std::string& s) {
        auto [it, inserted] = vocab_.try_emplace(s, next_id);
        if (inserted)
            ++next_id;
        return it->second;
    };
    std::vector<WordState> words;
    words.reserve(word_freq.size());
    for (const auto& [w, freq] : word_freq) {
        WordState ws;
        ws.count = freq;
        std::size_t pos = 0;
        bool first = true;
        while (pos < w.size()) {
            unsigned char c0 = static_cast<unsigned char>(w[pos]);
            std::size_t cp_len;
            if (c0 < 0x80)
                cp_len = 1;
            else if ((c0 >> 5) == 0b110)
                cp_len = 2;
            else if ((c0 >> 4) == 0b1110)
                cp_len = 3;
            else if ((c0 >> 3) == 0b11110)
                cp_len = 4;
            else
                cp_len = 1;
            if (pos + cp_len > w.size())
                cp_len = w.size() - pos;
            std::string ch = w.substr(pos, cp_len);
            std::string tok = first ? ch : continuing_prefix_ + ch;
            add_token(tok);
            ws.symbols.push_back(std::move(tok));
            first = false;
            pos += cp_len;
        }
        words.push_back(std::move(ws));
    }
    // 3. Greedy pair merging.
    while (vocab_.size() < target_vocab_size) {
        std::unordered_map<std::string, std::uint64_t> pair_freq;
        for (const auto& ws : words) {
            for (std::size_t k = 0; k + 1 < ws.symbols.size(); ++k) {
                std::string key = ws.symbols[k] + '\x01' + ws.symbols[k + 1];
                pair_freq[key] += ws.count;
            }
        }
        if (pair_freq.empty())
            break;
        auto best_it = pair_freq.begin();
        for (auto it = pair_freq.begin(); it != pair_freq.end(); ++it) {
            if (it->second > best_it->second ||
                (it->second == best_it->second && it->first < best_it->first)) {
                best_it = it;
            }
        }
        if (best_it->second < 2)
            break;
        std::string key = best_it->first;
        auto sep = key.find('\x01');
        std::string left = key.substr(0, sep);
        std::string right = key.substr(sep + 1);
        // Form the merged surface.  Strip the right half's "##"
        // before concatenation (the merged token keeps "##" iff the
        // LEFT half had one).
        std::string right_core = right;
        if (right.size() >= continuing_prefix_.size() &&
            right.compare(0, continuing_prefix_.size(), continuing_prefix_) == 0) {
            right_core = right.substr(continuing_prefix_.size());
        }
        std::string merged = left + right_core;
        add_token(merged);
        // Rewrite words to apply the merge.
        for (auto& ws : words) {
            std::vector<std::string> next;
            next.reserve(ws.symbols.size());
            std::size_t k = 0;
            while (k < ws.symbols.size()) {
                if (k + 1 < ws.symbols.size() && ws.symbols[k] == left &&
                    ws.symbols[k + 1] == right) {
                    next.push_back(merged);
                    k += 2;
                } else {
                    next.push_back(ws.symbols[k]);
                    ++k;
                }
            }
            ws.symbols = std::move(next);
        }
    }
    rebuild_tables_();
}

}  // namespace lucid::utils::tokenizer

File: lucid/utils/tokenizer/_wordpiece.py
"""WordPieceTokenizer — the algorithm BERT/DistilBERT/RoFormer use.

Encode algorithm (greedy longest-match)
---------------------------------------
For each pre-tokenized word:

1. If the whole word is in vocab, emit its id.
2. Otherwise, walk the word left-to-right finding the longest
   prefix that IS in vocab, emit its id, then recurse on the
   remainder — but the remainder must be looked up with the
   continuation prefix (``"##"`` by default) prepended to mark it
   as a non-initial subword.
3. If no valid prefix exists at any position, emit the UNK token
   (typically ``[UNK]``) for the entire word.

On-disk format
--------------
HF-compatible ``vocab.txt`` — one token per line, id = line index.
Continuation pieces start with ``##``.  Loads/saves any published
BERT-family checkpoint without modification.

Training
--------
Greedy frequency-based BPE-style training (matches HF
``WordPieceTrainer`` semantics; faster than the full log-likelihood
training in the original paper).  Standard pre-tokenizer chain is
:class:`BertNormalizer` + :class:`WhitespacePunctuationSplit` —
configure via the constructor if you need a different one.
"""

import os
from typing import Iterable

from lucid._C import engine as _C_engine

from lucid.utils.tokenizer._base import SpecialTokens, Tokenizer
from lucid.utils.tokenizer._lookup_common import (
    _load_special_tokens_map,
    _load_vocab_txt,
    _save_vocab_txt,
)
from lucid.utils.tokenizer._normalizers import BertNormalizer, Normalizer
from lucid.utils.tokenizer._pre_tokenizers import (
    PreTokenizer,
    WhitespacePunctuationSplit,
)


class _WordPieceCommonMixin:
    """Shared algorithm + post-processing for both WordPiece flavours."""

    _normalizer: Normalizer | None
    _pre_tokenizer: PreTokenizer
    _continuing_prefix: str

    def _prepare_words(self, text: str) -> list[str]:
        """Apply normalizer + pre-tokenizer; return word strings."""
        if self._normalizer is not None:
            text = self._normalizer(text)
        return [chunk for chunk, _ in self._pre_tokenizer(text)]

    def _decode_join(self, ids: list[int], id_to_token: dict[int, str]) -> str:
        """Standard BERT-style decode: join tokens; strip ``##`` from
        continuation pieces; emit a space before each non-continuation."""
        out: list[str] = []
        first = True
        for i in ids:
            tok = id_to_token.get(i)
            if tok is None:
                continue
            if tok.startswith(self._continuing_prefix):
                out.append(tok[len(self._continuing_prefix) :])
            else:
                if not first:
                    out.append(" ")
                out.append(tok)
            first = False
        return "".join(out)


# ── Pure-Python WordPiece ───────────────────────────────────────────


class WordPieceTokenizer(_WordPieceCommonMixin, Tokenizer):
    r"""Reference (pure-Python) WordPiece tokenizer.

    Parameters
    ----------
    vocab : dict[str, int]
        BERT-style vocab.  Continuation tokens prefixed with ``##``.
    unk_token : str, default "[UNK]"
        Token emitted when no valid longest-match prefix can be
        found.  Must be present in ``vocab`` for it to actually
        emit an id (otherwise emits silently nothing).
    continuing_prefix : str, default "##"
        Prefix for non-initial subwords.
    max_chars_per_word : int, default 100
        Words longer than this are immediately emitted as UNK
        (matches BERT's ``BasicTokenizer`` cap).
    normalizer : Normalizer, optional
        Default :class:`BertNormalizer`.
    pre_tokenizer : PreTokenizer, optional
        Default :class:`WhitespacePunctuationSplit`.
    special_tokens : SpecialTokens, optional
        Defaults to a registry with ``unk=unk_token`` so the
        tokenizer behaves correctly out of the box.
    """

    def __init__(
        self,
        vocab: dict[str, int],
        *,
        unk_token: str = "[UNK]",
        continuing_prefix: str = "##",
        max_chars_per_word: int = 100,
        normalizer: Normalizer | None = None,
        pre_tokenizer: PreTokenizer | None = None,
        special_tokens: SpecialTokens | None = None,
    ) -> None:
        self._vocab: dict[str, int] = dict(vocab)
        self._id_to_token: dict[int, str] = {v: k for k, v in self._vocab.items()}
        self._unk_token = unk_token
        self._continuing_prefix = continuing_prefix
        self._max_chars_per_word = max_chars_per_word
        self._normalizer = normalizer if normalizer is not None else BertNormalizer()
        self._pre_tokenizer = (
            pre_tokenizer if pre_tokenizer is not None else WhitespacePunctuationSplit()
        )
        # Default special-tokens registry mirrors BERT's set when
        # the caller doesn't supply one.
        if special_tokens is None:
            special_tokens = SpecialTokens(unk=unk_token)
        super().__init__(special_tokens=special_tokens)

    @property
    def vocab_size(self) -> int:
        return len(self._vocab)

    @property
    def algo(self) -> str:
        return "wordpiece"

    def get_vocab(self) -> dict[str, int]:
        return dict(self._vocab)

    def id_to_token(self, token_id: int) -> str | None:
        return self._id_to_token.get(token_id)

    def _encode_word(self, word: str) -> list[int]:
        """Apply greedy longest-match to one pre-tokenized word."""
        # Whole-word shortcut.
        if word in self._vocab:
            return [self._vocab[word]]
        if len(word) > self._max_chars_per_word:
            unk_id = self._vocab.get(self._unk_token)
            return [unk_id] if unk_id is not None else []
        # Greedy longest-match left-to-right.
        ids: list[int] = []
        start = 0
        while start < len(word):
            match_id = None
            match_end = start
            for end in range(len(word), start, -1):
                sub = word[start:end]
                if start > 0:
                    sub = self._continuing_prefix + sub
                tid = self._vocab.get(sub)
                if tid is not None:
                    match_id = tid
                    match_end = end
                    break
            if match_id is None:
                # No prefix matched at this position → whole word is UNK.
                unk_id = self._vocab.get(self._unk_token)
                return [unk_id] if unk_id is not None else []
            ids.append(match_id)
            start = match_end
        return ids

    def _encode_one(self, text: str) -> list[int]:
        out: list[int] = []
        for word in self._prepare_words(text):
            out.extend(self._encode_word(word))
        return out

    def _decode_one(self, ids: list[int]) -> str:
        return self._decode_join(ids, self._id_to_token)

    def train(
        self,
        corpus: Iterable[str],
        *,
        vocab_size: int = 30_000,
    ) -> None:
        """Greedy frequency-based WordPiece training (BPE-like).

        For each iteration, count adjacent symbol-pair frequencies
        across all words + merge the most frequent.  Same
        deterministic tie-break (lexicographic key order) as BPE.

        The result is a BERT-compatible vocab with ``##``
        continuation markers and the configured UNK token.
        """
        # Pre-tokenize + count word frequencies.
        word_freq: dict[str, int] = {}
        for doc in corpus:
            for w in self._prepare_words(doc):
                word_freq[w] = word_freq.get(w, 0) + 1
        # Seed vocab with the UNK token (id 0 reserved for it) + every
        # distinct codepoint that appears in any word (continuation
        # forms get ``##`` prefix).
        new_vocab: dict[str, int] = {self._unk_token: 0}
        next_id = 1
        # Per-word symbol sequence (head symbol is plain, tail symbols
        # carry the continuation prefix).
        words: list[tuple[list[str], int]] = []
        for w, freq in word_freq.items():
            syms: list[str] = []
            for i, ch in enumerate(w):
                tok = ch if i == 0 else self._continuing_prefix + ch
                if tok not in new_vocab:
                    new_vocab[tok] = next_id
                    next_id += 1
                syms.append(tok)
            words.append((syms, freq))
        # Greedy pair merging.
        while len(new_vocab) < vocab_size:
            pair_freq: dict[tuple[str, str], int] = {}
            for syms, freq in words:
                for k in range(len(syms) - 1):
                    pair_freq[(syms[k], syms[k + 1])] = (
                        pair_freq.get((syms[k], syms[k + 1]), 0) + freq
                    )
            if not pair_freq:
                break
            best_pair = max(pair_freq.items(), key=lambda kv: (kv[1], kv[0]))
            if best_pair[1] < 2:
                break
            left, right = best_pair[0]
            right_core = (
                right[len(self._continuing_prefix) :]
                if right.startswith(self._continuing_prefix)
                else right
            )
            merged = left + right_core
            new_vocab[merged] = next_id
            next_id += 1
            # Rewrite all word symbol sequences.
            new_words: list[tuple[list[str], int]] = []
            for syms, freq in words:
                out: list[str] = []
                k = 0
                while k < len(syms):
                    if k + 1 < len(syms) and syms[k] == left and syms[k + 1] == right:
                        out.append(merged)
                        k += 2
                    else:
                        out.append(syms[k])
                        k += 1
                new_words.append((out, freq))
            words = new_words
        self._vocab = new_vocab
        self._id_to_token = {v: k for k, v in new_vocab.items()}
        self._refresh_special_ids()

    def save(self, directory: str) -> None:
        os.makedirs(directory, exist_ok=True)
        _save_vocab_txt(self._vocab, os.path.join(directory, "vocab.txt"))
        super().save(directory)

    def _save_extras(self) -> dict[str, object]:
        return {
            "model": {
                "type": "WordPiece",
                "vocab": self._vocab,
                "unk_token": self._unk_token,
                "continuing_prefix": self._continuing_prefix,
                "max_chars_per_word": self._max_chars_per_word,
            }
        }

    @classmethod
    def from_file(
        cls,
        directory: str,
        *,
        unk_token: str = "[UNK]",
        continuing_prefix: str = "##",
        max_chars_per_word: int = 100,
        normalizer: Normalizer | None = None,
        pre_tokenizer: PreTokenizer | None = None,
        special_tokens: SpecialTokens | None = None,
    ) -> "WordPieceTokenizer":
        """Load from a BERT-style directory containing ``vocab.txt``."""
        vocab_path = os.path.join(directory, "vocab.txt")
        if not os.path.isfile(vocab_path):
            raise FileNotFoundError(
                f"WordPieceTokenizer.from_file: vocab.txt not found in " f"{directory}"
            )
        vocab = _load_vocab_txt(vocab_path)
        st = special_tokens or _load_special_tokens_map(directory)
        return cls(
            vocab,
            unk_token=unk_token,
            continuing_prefix=continuing_prefix,
            max_chars_per_word=max_chars_per_word,
            normalizer=normalizer,
            pre_tokenizer=pre_tokenizer,
            special_tokens=st,
        )

    from_pretrained = from_file


# ── Fast (C++-backed) WordPiece ─────────────────────────────────────


class WordPieceTokenizerFast(_WordPieceCommonMixin, Tokenizer):
    """C++-backed WordPiece tokenizer.  See :class:`WordPieceTokenizer`.

    The greedy longest-match encode + training loop run in C++ via
    :class:`lucid._C.engine.utils.tokenizer.WordPiece`.  The Python
    side handles normalisation + pre-tokenization (so the same
    pre-processing chain works for both flavours).
    """

    def __init__(
        self,
        vocab: dict[str, int],
        *,
        unk_token: str = "[UNK]",
        continuing_prefix: str = "##",
        max_chars_per_word: int = 100,
        normalizer: Normalizer | None = None,
        pre_tokenizer: PreTokenizer | None = None,
        special_tokens: SpecialTokens | None = None,
    ) -> None:
        self._vocab: dict[str, int] = dict(vocab)
        self._id_to_token: dict[int, str] = {v: k for k, v in self._vocab.items()}
        self._unk_token = unk_token
        self._continuing_prefix = continuing_prefix
        self._max_chars_per_word = max_chars_per_word
        self._cpp = _C_engine.utils.tokenizer.WordPiece(
            self._vocab,
            unk_token,
            continuing_prefix,
            max_chars_per_word,
        )
        self._normalizer = normalizer if normalizer is not None else BertNormalizer()
        self._pre_tokenizer = (
            pre_tokenizer if pre_tokenizer is not None else WhitespacePunctuationSplit()
        )
        if special_tokens is None:
            special_tokens = SpecialTokens(unk=unk_token)
        super().__init__(special_tokens=special_tokens)

    @property
    def vocab_size(self) -> int:
        return self._cpp.vocab_size()

    @property
    def algo(self) -> str:
        return "wordpiece"

    def get_vocab(self) -> dict[str, int]:
        return dict(self._vocab)

    def id_to_token(self, token_id: int) -> str | None:
        return self._id_to_token.get(token_id)

    def _encode_one(self, text: str) -> list[int]:
        """Normalize + pre-tokenize in Python; encode each word in C++."""
        out: list[int] = []
        for word in self._prepare_words(text):
            out.extend(self._cpp.encode(word))
        return out

    def _decode_one(self, ids: list[int]) -> str:
        return self._decode_join(ids, self._id_to_token)

    def train(
        self,
        corpus: Iterable[str],
        *,
        vocab_size: int = 30_000,
    ) -> None:
        """Train in C++ via :meth:`WordPiece.train`.

        Pre-encode the corpus through the configured normalizer +
        pre-tokenizer in Python so the C++ training loop sees
        already-normalized, whitespace-separated words (its internal
        splitter is also whitespace, so this composes correctly).
        """
        encoded_corpus: list[str] = []
        for doc in corpus:
            words = self._prepare_words(doc)
            encoded_corpus.append(" ".join(words))
        self._cpp.train(encoded_corpus, vocab_size)
        self._vocab = dict(self._cpp.get_vocab())
        self._id_to_token = {v: k for k, v in self._vocab.items()}
        self._refresh_special_ids()

    def save(self, directory: str) -> None:
        os.makedirs(directory, exist_ok=True)
        _save_vocab_txt(self._vocab, os.path.join(directory, "vocab.txt"))
        super().save(directory)

    def _save_extras(self) -> dict[str, object]:
        return {
            "model": {
                "type": "WordPiece",
                "vocab": self._vocab,
                "unk_token": self._unk_token,
                "continuing_prefix": self._continuing_prefix,
                "max_chars_per_word": self._max_chars_per_word,
            }
        }

    @classmethod
    def from_file(
        cls,
        directory: str,
        *,
        unk_token: str = "[UNK]",
        continuing_prefix: str = "##",
        max_chars_per_word: int = 100,
        normalizer: Normalizer | None = None,
        pre_tokenizer: PreTokenizer | None = None,
        special_tokens: SpecialTokens | None = None,
    ) -> "WordPieceTokenizerFast":
        vocab_path = os.path.join(directory, "vocab.txt")
        if not os.path.isfile(vocab_path):
            raise FileNotFoundError(
                f"WordPieceTokenizerFast.from_file: vocab.txt not found "
                f"in {directory}"
            )
        vocab = _load_vocab_txt(vocab_path)
        st = special_tokens or _load_special_tokens_map(directory)
        return cls(
            vocab,
            unk_token=unk_token,
            continuing_prefix=continuing_prefix,
            max_chars_per_word=max_chars_per_word,
            normalizer=normalizer,
            pre_tokenizer=pre_tokenizer,
            special_tokens=st,
        )

    from_pretrained = from_file
